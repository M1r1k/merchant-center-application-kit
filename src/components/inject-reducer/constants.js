// eslint-disable-next-line import/prefer-default-export
export const ACTIVATE_PLUGIN = 'ACTIVATE_PLUGIN';
