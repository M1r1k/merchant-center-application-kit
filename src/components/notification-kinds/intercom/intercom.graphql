mutation ChangeIntercomStatus($status: IntercomTrackingStatus!) {
  changeIntercomStatus(status: $status) {
    id
    tracking_intercom
  }
}
