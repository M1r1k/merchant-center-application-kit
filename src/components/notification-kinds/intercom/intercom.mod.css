.underline {
  composes: underline from '@commercetools-local/core/_styles/commons.mod.css';
}

.link {
  composes: bold link from '@commercetools-local/core/_styles/commons.mod.css';
}
