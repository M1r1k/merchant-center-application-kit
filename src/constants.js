// eslint-disable-next-line import/prefer-default-export
export const USER_PROFILE_FORM_NAME =
  'packages-shared/app-shell/USER_PROFILE_FORM_NAME';
