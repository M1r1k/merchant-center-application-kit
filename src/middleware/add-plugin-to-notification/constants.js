export const __LOCAL = 'LOCAL';
export const __GLOBAL = 'GLOBAL';
