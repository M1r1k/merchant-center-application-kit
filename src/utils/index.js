export { default as selectProjectKey } from './select-project-key';
export { default as selectUserId } from './select-user-id';
export { default as getCorrelationId } from './get-correlation-id';
