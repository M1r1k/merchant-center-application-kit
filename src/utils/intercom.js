const createIntercomUserDataForUser = user => ({
  app_id: window.app.tracking.intercomId,
  user_id: user.id,
  // This enables secure mode. Once it's enabled, it stays on!
  user_hash: user.intercom_user_hash,
  name: `${user.firstName} ${user.lastName}`,
  email: user.email,
  signed_up_at: user.createdAt, // either a Unix timestamp or ISO 8601
  language: user.language,
  mc_location: window.app.location,
  ...(user.organization
    ? {
        company: {
          id: user.organization.id,
          name: user.organization.name,
          created_at: user.organization.createdAt,
          teams_count: user.organization.teamsCount,
        },
      }
    : {}),
});

export const boot = user => {
  if (!window.Intercom || !window.app.tracking.intercomId) {
    return;
  }
  // From the Intercom SPA Docs
  // You boot the user as normal:
  // window.Intercom('boot', {app_id: APP_ID, email: 'someuser@example.com', apples: 1});
  // https://docs.intercom.com/install-on-your-product-or-site/other-ways-to-get-started/integrate-intercom-in-a-single-page-app
  window.Intercom('boot', createIntercomUserDataForUser(user));
};

export const shutdown = () => {
  if (!window.Intercom || !window.app.tracking.intercomId) {
    return;
  }
  // From the Intercom SPA Docs
  // When your user logs out of Intercom (or is automatically logged out by your
  // app), call Intercom('shutdown');  from our JavaScript API, to end the
  // Intercom session and clear the cookie.
  // https://docs.intercom.com/install-on-your-product-or-site/other-ways-to-get-started/integrate-intercom-in-a-single-page-app
  window.Intercom('shutdown');
};

export const updateUser = user => {
  if (!window.Intercom || !window.app.tracking.intercomId) {
    return;
  }
  // From the Intercom SPA Docs
  // If the user suddenly has two apples, you would call:
  // window.Intercom('update', {email: 'someuser@example.com', apples: 2});
  // https://docs.intercom.com/install-on-your-product-or-site/other-ways-to-get-started/integrate-intercom-in-a-single-page-app
  window.Intercom('update', createIntercomUserDataForUser(user));
};

export const changePage = () => {
  if (!window.Intercom || !window.app.tracking.intercomId) {
    return;
  }
  // From the Intercom SPA Docs
  // But if a user has just caused a â€œpage" change, you would call:
  // window.Intercom('update');
  // https://docs.intercom.com/install-on-your-product-or-site/other-ways-to-get-started/integrate-intercom-in-a-single-page-app
  window.Intercom('update');
};
