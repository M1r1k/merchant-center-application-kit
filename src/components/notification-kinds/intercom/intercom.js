import PropTypes from 'prop-types';
import React from 'react';
import { defineMessages, FormattedMessage } from 'react-intl';
import { compose } from 'recompose';
import { graphql } from 'react-apollo';
import logger from '@commercetools-local/utils/logger';
import { GRAPHQL_TARGETS, DOMAINS } from '@commercetools-local/constants';
import Notification from '@commercetools-local/core/components/notification';
import { injectConfiguration } from '@commercetools-local/core/components/configuration';
import { INTERCOM_TRACKING_STATUS } from '../../../constants';
import ChangeIntercomStatus from './intercom.graphql';
import styles from './intercom.mod.css';

const messages = defineMessages({
  title: {
    id: 'GlobalNotification.title',
    description:
      'User logs in for first time and needs to opt-in/opt-out of intercom tracking / Appears for first-time users',
    defaultMessage:
      'In order to constantly improve the Merchant Center experience, we use the {linkToIntercom} analytics services. You can {link} of having this information collected anytime.',
  },
  optOutLink: {
    id: 'GlobalNotification.optOutLink',
    description: 'Text for the "opt-out" action.',
    defaultMessage: 'opt-out',
  },
});

export class IntercomNotification extends React.PureComponent {
  static displayName = 'IntercomNotification';

  static propTypes = {
    dismiss: PropTypes.func.isRequired,
    notification: PropTypes.shape({
      id: PropTypes.number.isRequired,
      kind: PropTypes.oneOf(['intercom']).isRequired,
      domain: PropTypes.oneOf([DOMAINS.GLOBAL]).isRequired,
    }),
    showUnexpectedErrorNotification: PropTypes.func.isRequired,
    // Injected
    changeIntercomStatus: PropTypes.func.isRequired,
    environmentName: PropTypes.string.isRequired,
  };

  handleMutationError = error => {
    // On production we send the errors to Sentry.
    if (this.props.environmentName !== 'production')
      logger.error(error, error.stack);
    // Show an error message
    this.props.showUnexpectedErrorNotification({ error });
  };

  handleOptOutLinkClick = () => {
    this.props
      .changeIntercomStatus({
        variables: {
          target: GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND,
          status: INTERCOM_TRACKING_STATUS.inactive,
        },
      })
      .then(() => {
        this.props.dismiss();
        // no need to shut down intercom manually since the users intercom
        // tracking status was pending before this notification was shown
        // intercom never got booted in the first place
      })
      .catch(error => this.handleMutationError(error));
  };

  handleCloseNotification = () => {
    this.props
      .changeIntercomStatus({
        variables: {
          target: GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND,
          status: INTERCOM_TRACKING_STATUS.active,
        },
      })
      .then(() => {
        this.props.dismiss();
      })
      .catch(error => this.handleMutationError(error));
  };

  render() {
    return (
      <Notification
        type="info"
        onCloseClick={this.handleCloseNotification}
        domain={this.props.notification.domain}
      >
        <FormattedMessage
          {...messages.title}
          values={{
            linkToIntercom: (
              <a
                className={styles.underline}
                href="https://www.intercom.io"
                target="_blank"
                rel="noreferrer noopener"
              >
                {'Intercom'}
              </a>
            ),
            link: (
              <span
                className={styles.link}
                onClick={this.handleOptOutLinkClick}
              >
                <FormattedMessage {...messages.optOutLink} />
              </span>
            ),
          }}
        />
      </Notification>
    );
  }
}

export default compose(
  graphql(ChangeIntercomStatus, { name: 'changeIntercomStatus' }),
  injectConfiguration(['env'], 'environmentName')
)(IntercomNotification);
