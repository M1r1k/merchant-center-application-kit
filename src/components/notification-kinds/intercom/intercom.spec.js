import React from 'react';
import { shallow } from 'enzyme';
import { GraphQlTargets, DOMAINS } from '@commercetools-local/constants';
import { INTERCOM_TRACKING_STATUS } from '../../../constants';
import * as intercom from '../../../utils/intercom';
import { IntercomNotification } from './intercom';

jest.mock('../../../utils/intercom');

const createTestProps = custom => ({
  dismiss: jest.fn(),
  notification: {
    id: 1,
    kind: 'intercom',
    domain: DOMAINS.GLOBAL,
  },
  changeIntercomStatus: jest.fn(),
  showUnexpectedErrorNotification: jest.fn(),
  environmentName: 'test',
  ...custom,
});

describe('rendering', () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<IntercomNotification {...props} />);
  });
  it('should render a Notification wrapper', () => {
    expect(wrapper).toRender('Notification');
  });
  it('should render a Message with title', () => {
    expect(wrapper.find('FormattedMessage')).toHaveProp(
      'id',
      'GlobalNotification.title'
    );
  });
  describe('values for FormattedMessage', () => {
    let valueWrapper;
    describe('linkToIntercom', () => {
      beforeEach(() => {
        valueWrapper = shallow(
          wrapper.find('FormattedMessage').prop('values').linkToIntercom
        );
      });
      it('should match snapshot', () => {
        expect(valueWrapper).toMatchSnapshot();
      });
    });
    describe('link', () => {
      beforeEach(() => {
        valueWrapper = shallow(
          wrapper.find('FormattedMessage').prop('values').link
        );
      });
      it('should match snapshot', () => {
        expect(valueWrapper).toMatchSnapshot();
      });
    });
  });
});

describe('interactions', () => {
  let props;
  let wrapper;
  describe('handleMutationError', () => {
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<IntercomNotification {...props} />);
      return wrapper.instance().handleMutationError(new Error('Oops'));
    });
    it('should dispatch error notification', () => {
      expect(props.showUnexpectedErrorNotification).toHaveBeenCalledTimes(1);
      expect(props.showUnexpectedErrorNotification).toHaveBeenCalledWith({
        error: new Error('Oops'),
      });
    });
  });
  describe('handleOptOutLinkClick', () => {
    beforeEach(() => {
      intercom.shutdown.mockClear();
      props = createTestProps({
        changeIntercomStatus: jest.fn(() => Promise.resolve()),
      });
      wrapper = shallow(<IntercomNotification {...props} />);
      return wrapper.instance().handleOptOutLinkClick();
    });
    it('should trigger mutation', () => {
      expect(props.changeIntercomStatus).toHaveBeenCalledTimes(1);
    });
    it('should trigger mutation passing variable "mc" target', () => {
      expect(props.changeIntercomStatus).toHaveBeenCalledWith({
        variables: expect.objectContaining({
          target: GraphQlTargets.MerchantCenterBackend,
        }),
      });
    });
    it('should trigger mutation passing variable "INACTIVE" status', () => {
      expect(props.changeIntercomStatus).toHaveBeenCalledWith({
        variables: expect.objectContaining({
          status: INTERCOM_TRACKING_STATUS.inactive,
        }),
      });
    });
    it('should dismiss notification', () => {
      expect(props.dismiss).toHaveBeenCalledTimes(1);
    });
  });
  describe('handleCloseNotification', () => {
    beforeEach(() => {
      intercom.changePage.mockClear();
      props = createTestProps({
        changeIntercomStatus: jest.fn(() => Promise.resolve()),
      });
      wrapper = shallow(<IntercomNotification {...props} />);
      return wrapper.instance().handleCloseNotification();
    });
    it('should trigger mutation', () => {
      expect(props.changeIntercomStatus).toHaveBeenCalledTimes(1);
    });
    it('should trigger mutation passing variable "mc" target', () => {
      expect(props.changeIntercomStatus).toHaveBeenCalledWith({
        variables: expect.objectContaining({
          target: GraphQlTargets.MerchantCenterBackend,
        }),
      });
    });
    it('should trigger mutation passing variable "ACTIVE" status', () => {
      expect(props.changeIntercomStatus).toHaveBeenCalledWith({
        variables: expect.objectContaining({
          status: INTERCOM_TRACKING_STATUS.active,
        }),
      });
    });
    it('should dismiss notification', () => {
      expect(props.dismiss).toHaveBeenCalledTimes(1);
    });
  });
});
