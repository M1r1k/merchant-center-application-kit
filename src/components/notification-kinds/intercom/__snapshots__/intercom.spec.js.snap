// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering values for FormattedMessage link should match snapshot 1`] = `
<span
  className="link"
  onClick={[Function]}
>
  <FormattedMessage
    defaultMessage="opt-out"
    description="Text for the \\"opt-out\\" action."
    id="GlobalNotification.optOutLink"
    values={Object {}}
  />
</span>
`;

exports[`rendering values for FormattedMessage linkToIntercom should match snapshot 1`] = `
<a
  className="underline"
  href="https://www.intercom.io"
  rel="noreferrer noopener"
  target="_blank"
>
  Intercom
</a>
`;
